lib/spack/docs/tutorial_modules.rst

```diff
@@ -1,52 +1,250 @@
+    module files. It's a drop-in replacement of Environment Modules and works with
+    both of the module file formats generated by Spack.
+    Despite being fully compatible with Environment Modules there are many features that
+    are unique to Lmod. These features are either
+    `targeted to safety <http://lmod.readthedocs.io/en/latest/010_user.html#safety-features>`_
+    or meant to
+    `extend the module system functionality <http://lmod.readthedocs.io/en/latest/010_user.html#module-hierarchy>`_.
+
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How do we generate module files?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Before  we dive into the hands-on it's worth spending a couple of words to explain how
+module files are generated by Spack. The following diagram provides a high-level view
+of the process:
+
+
+.. image:: module_file_generation.*
+
+The red dashed line above represents Spack's boundaries, the blue one Spack's vendored
+dependencies. In short, module files are generated combining:
+
+  * the configuration details in ``config.yaml`` and ``modules.yaml``
+  * the information contained in Spack packages (and processed by the module subpackage)
+  * a set of `Jinja2 <http://jinja.pocoo.org/docs/2.9/>`_ templates
+
+with the vendored template engine to "stamp out" each particular module file. As Spack serves
+very diverse needs this process has many points of customization, and we'll explore most of
+them in the next sections.
 
 .. _module_file_tutorial_prerequisites:
 
--------------
-Prerequisites
--------------
+-----------------------
+Set up for the tutorial
+-----------------------
 
-Before proceeding further ensure:
+In order to showcase the capabilities of Spack's module file generation, we need to
+prepare a representative set of software to work with. This set includes different
+flavors of the same packages installed alongside each other and some
+:ref:`external packages <sec-external-packages>` (including MPI implementations).
 
-- you have LMod or Environment Modules available
-- have :ref:`shell support <shell-support>` activated in Spack
+We are not doing this just to make our life harder, and
+see how we can still get out of the situation thanks to Spack. The set-up is intended
+to mimic closely what happens on real HPC clusters, where some dependencies are provided
+by vendors (and are better not to be built again by Spack).
 
-If you need to install Lmod or Environment module you can refer
-to the documentation :ref:`here <InstallEnvironmentModules>`.
+The best way to follow along is to use a docker image (see
+:ref:`module_file_tutorial_use_docker`). If for any reason you prefer to work
+locally you should read instead :ref:`module_file_tutorial_work_locally` (but be aware
+that all the snippets afterwards come from the docker image, so your mileage may vary).
 
+.. _module_file_tutorial_use_docker:
 
 ^^^^^^^^^^^^^^^^^^
-Add a new compiler
+Use a Docker image
```

Comment:
You're right. I gave it for granted that the purpose of the Docker image was to avoid installing software. It's explicit now. Let me know if the new phrasing is clear enough.